libs/header/sync/sync.go

```diff
@@ -276,25 +276,25 @@ func (s *Syncer[H]) findHeaders(ctx context.Context, from, to uint64) ([]H, erro
 	if amount > s.Params.MaxRequestSize {
 		to, amount = from+s.Params.MaxRequestSize, s.Params.MaxRequestSize
 	}
-
 	out := make([]H, 0, amount)
-	for from < to {
+	for from <= to {
 		// if we have some range cached - use it
 		r, ok := s.pending.FirstRangeWithin(from, to)
 		if !ok {
-			hs, err := s.exchange.GetRangeByHeight(ctx, from, amount)
+			hs, err := s.exchange.GetRangeByHeight(ctx, from, to-from+1)
 			return append(out, hs...), err
 		}
 
-		// first, request everything between from and start of the found range
-		hs, err := s.exchange.GetRangeByHeight(ctx, from, r.start-from)
-		if err != nil {
-			return nil, err
+		if r.start > from {
```

Comment:
Now that I have come back to this code again. I recalled that I made this specifically to have a case with 0, as I expected it to be handled by the implementation. Basically hide the if inside GetRangeByHeight.

I think all implementations must have a sanity check for when zero GetRangeByHeight is as you initially did, independently of what we do in Syncer. The sanity check should either return an error when the 0 range is requested or return an empty slice as a successful result for requested range 0.

If we go with the successful results for request range 0, there is no big reason to have the `if r.start > from` as anyway, we will get zero len slice which will be noop. But if we go with returning the error when 0 range is requested, we would have to keep the if. 

Pls choice the option you are most comfortable with regarding the if, but we still need to have sanity checks in Getter impls

Reply:
I'm ok with returning an empty slice(actually I've done this already, but then decided to rework it).
So, do you want me to add sanity checks in all getters in scope of this PR? Or should I do it only for p2p.Exchange(all others in subsequent pr)?

Reply:
I am fine with doing such a check in this PR